### DISCUSSION TITLE
February 4 Dev Session

### DISCUSSION THREAD

**Discussion by tpike3** [mesa/mesa#1588] [General]
When: 4 Feb --- 2000 UTC+8/1300 CET/1200 GMT/0700 EDT

Where: [via Matrix](https://matrix.to/#/#mesa-dev:matrix.org)

Proposed Agenda:

- GSoC
- quick update on tutorial
- mesa-geo
- issues for next release
- issues for Mesa 2.0

@jackiekazil @rht @Corvince @wang-boyu @Tortar @aiwhoo @nunezco2 @EwoutH @Logende Please feel free to add or takeaway from this agenda

**Comment by rht** [comment]
@Tortar and I had made some discoveries with speeding up space.py and time.py with Cython.
I want to discuss on whether we should:
- release the Cythonized code as a separate package, advertise to users. If they use it much more often than the pure Python version, then we integrate to core Mesa
- replace the current Python code with the Cythonized code, in Mesa 2.0. They are not that different anyway. Additionally, this reduces maintenance overhead of having to maintain 2 very similar code.

The space benchmark is impressive
```
timings cython

cython grid init 76.222 μs --> speedup 8.85
cython grid get_neighborhood 0.154 μs --> speedup 1.63
cython grid get_cell_list_contents 0.217 μs --> speedup 10.72
cython grid get_neighbors 0.337 μs --> speedup 7.26
cython grid out_of_bounds 0.034 μs --> speedup 4.16
cython grid is_cell_empty 0.034 μs --> speedup 4.49
cython grid move_to_empty 3.005 μs --> speedup 1.74
cython grid remove_agent 0.039 μs --> speedup 1.97
cython grid place_agent 0.129 μs --> speedup 4.57
cython grid torus_adj 0.061 μs --> speedup 4.27
cython grid iter_cell_list_contents 0.637 μs --> speedup 3.36
cython grid iter_neighbors 0.775 μs --> speedup 2.92
cython grid coord_iter 501.780 μs --> speedup 1.63
cython grid iter 143.104 μs --> speedup 1.07
```
And here is the benchmark for `BaseScheduler` and `RandomActivation`:
```
default 144.252 μs
default shuffled 454.117 μs
cython python dict 92.468 μs
  1.56x
cython python dict, shuffled 110.337 μs
  4.12x
cython map, cython agent 16.105 μs
  8.96x
cython map, cython agent, cython shuffled 32.478 μs
  13.98x
```

**Comment by wang-boyu** [comment]
Some issues/PRs to discuss for Mesa-Geo:
- https://github.com/projectmesa/mesa-geo/issues/46
- https://github.com/projectmesa/mesa-examples/pull/17
- https://github.com/projectmesa/mesa-geo/issues/131

**Comment by tpike3** [comment]
Apologies for the delay on this but here are the notes from the 4 FEB Dev Session- 

Regarding Mesa Development - 

We discussed the work @Tortar and @rht did to cythonize Mesa to improve performance ([see #rht comments above](https://github.com/projectmesa/mesa/discussions/1588#discussioncomment-4827410). 
- We discussed integrating into mesa deployment which can be done is we update the CI/CD pipeline and specify the wheel as a dependency. 
- There was general agreement if we switch to a cython based backend (looking at doing it for Mesa 2.0) that we should have a cython version and non-cython version as it would be too difficult two maintain versions.   Although there is some concerns about the impact of user experience. 

Regarding Mesa-Geo:
- we discussed use of cookie cutter for mesa geo
- discussed mesa-geo examples and attribution of base maps
- discussed the size of the mesa-geo particularly in relation to the examples. 

Finally we updated milestones so we can do release of Mesa 1.2 this month. 

Through NUMFOCUS we intend to participate in the [Google Summer of Code Projects](https://github.com/projectmesa/mesa/wiki/Projects-for-Google-Summer-of-Code-2023). We prioritized options and refined some of the projects. In addition, since the dev session we got some feedback and added some desired features. 

We sent out a [NUMFOCUS newsletter](https://numfocus.salsalabs.org/numfocus__newsletter_feb2023?wvpId=dcc268c0-a7fa-4e9a-a763-f26ea9dff565) and sent out an update to group of Universities that are employing Mesa in the curriculum.
